docs/source/coding_guidelines.md

```diff
@@ -0,0 +1,213 @@
+
+For non-trivial changes, we recommend to work with a team member to create a design for your work.
+Depending on the size of the task at hand, a design can be anywhere between a short descrption of
+your plan on Github, or a long document where you list your design options and invite feedback.
+
+Your design partner will later also be well positioned to review your PR because they are familiar with your plan.
+
+## Setting up your dev environment
+
+Please see the detailed instructions [here](developers.md).
+
+## Coding Guidelines
+
+### Naming
+
+Please follow the general Python rules for naming:
+
+* Variables, function and method names should use `snake_case` (lower case with under scores separating words)
+* Class names should follow `PascalCase` (first letter of each word capitalized).
+
+In addition:
+
+* To improve readability, functions or methods that return boolean values should follow a `is_...`, `has_...`, `use...` pattern, like `is_status_ok` instead of `ok`.
+
+### Static Analysis and Linting
+
+We use `flake8` as a linter, and `mypy` and pyright for static typechecking. Both tools run as part of the PR build, and must run without errors for a contribution to be accepted. `mypy` requires that all functions and methods carry type annotations, see [mypy documentation](https://mypy.readthedocs.io/en/latest/getting_started.html#function-signatures-and-dynamic-vs-static-typing)
+
+We highly recommend to run all those tools _before_ pushing the latest changes to a PR. If you have `make` installed, you can run both tools in one go via `make check` (from the repository root folder)
+
+### String Style
+
+End sentences in docstrings with a period:
+
+```python
+def method(self, arg: int) -> None:
+    """
+    Method description, followed by an period.
+    """
+```
+
+* End strings in error messages with a period.
+
+### Documentation
+
+We are using Sphinx for generating documentation. We recommend using a VSCode extension for auto-generating documentation templates, for example `njpwerner.autodocstring` (this extension is already pre-configured in VSCode's workspace settings file).
+
+Common mistakes when writing docstrings:
+
+* There must be a separating line between a function description and the documentation for its parameters.
+* In multi-line parameter descriptions, continuations on the next line must be indented.
+* Sphinx will merge the class description and the arguments of the constructor `__init__`. Hence, there is no need to
+  write any text in the constructor docstring other than the constructor arguments.
+* Use `>>>` to include code snippets. PyCharm will run intellisense on those to make authoring easier.
+* To generate the Sphinx documentation on your dev machine, run `make html` in the `./docs` folder, and then
+  open `./docs/build/html/index.html`
```

Comment:
If we `cd`'ed to `docs` to run `make html`, then `./docs/build/html/index.html` is not correct. Shall we instruct to first `cd` and then open `build/html/index.html`?